{"cells":[{"cell_type":"code","execution_count":1,"metadata":{},"outputs":[],"source":"import glob\nimport pandas as pd\n\nfrom biom import Table, load_table\nfrom skbio.stats.composition import closure\nfrom gemelli.rpca import rpca_table_processing\nfrom sklearn.model_selection import train_test_split\n\n#turn off warnings\nimport warnings\nwarnings.filterwarnings(\"ignore\")"},{"cell_type":"markdown","metadata":{},"source":"## Data pre-processing\n\nOnly need to run these blocks once"},{"cell_type":"code","execution_count":2,"metadata":{},"outputs":[],"source":"save_path = '../../data/case-studies/time-ihmp-uc/'"},{"cell_type":"markdown","metadata":{},"source":"### iHMP cohort"},{"cell_type":"code","execution_count":14,"metadata":{},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>Participant ID</th>\n","      <th>site_sub_coll</th>\n","      <th>data_type</th>\n","      <th>week_num</th>\n","      <th>date_of_receipt</th>\n","      <th>interval_days</th>\n","      <th>visit_num</th>\n","      <th>Research Project</th>\n","      <th>PDO Number</th>\n","      <th>GSSR IDs</th>\n","      <th>...</th>\n","      <th>Proteomics received at LBNL:</th>\n","      <th>Stool Sample ID:  Tube A  (EtOH)</th>\n","      <th>Sample ID: Tube B (No preservative)</th>\n","      <th>Tube A and B received at Broad:</th>\n","      <th>stool_id</th>\n","      <th>smoking status</th>\n","      <th>Number years smoked</th>\n","      <th>Age when started smoking</th>\n","      <th>How many cigarettes/cigars/etc. do you smoke per day?</th>\n","      <th>Testing</th>\n","    </tr>\n","    <tr>\n","      <th>External ID</th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>CSM79HP4</th>\n","      <td>C3022</td>\n","      <td>C3022C11</td>\n","      <td>metabolomics</td>\n","      <td>21.0</td>\n","      <td>2015-04-21</td>\n","      <td>20.0</td>\n","      <td>15</td>\n","      <td>ibdmdb</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>...</td>\n","      <td>No</td>\n","      <td>SM-79HP4</td>\n","      <td>SM-79HP5</td>\n","      <td>No</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>Train</td>\n","    </tr>\n","    <tr>\n","      <th>MSM9VZMM</th>\n","      <td>M2084</td>\n","      <td>M2084C1</td>\n","      <td>metabolomics</td>\n","      <td>0.0</td>\n","      <td>2015-08-10</td>\n","      <td>0.0</td>\n","      <td>4</td>\n","      <td>ibdmdb</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>...</td>\n","      <td>No</td>\n","      <td>SM-9VZMM</td>\n","      <td>SM-9VZMN</td>\n","      <td>No</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>Train</td>\n","    </tr>\n","    <tr>\n","      <th>MSM79HDE</th>\n","      <td>M2069</td>\n","      <td>M2069C5</td>\n","      <td>metabolomics</td>\n","      <td>8.0</td>\n","      <td>2015-06-02</td>\n","      <td>14.0</td>\n","      <td>8</td>\n","      <td>ibdmdb</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>...</td>\n","      <td>No</td>\n","      <td>SM-79HDE</td>\n","      <td>SM-79HDF</td>\n","      <td>No</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>Train</td>\n","    </tr>\n","    <tr>\n","      <th>PSM6XBRK</th>\n","      <td>P6009</td>\n","      <td>P6009C1</td>\n","      <td>metabolomics</td>\n","      <td>0.0</td>\n","      <td>2014-11-12</td>\n","      <td>0.0</td>\n","      <td>4</td>\n","      <td>ibdmdb</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>...</td>\n","      <td>No</td>\n","      <td>SM-6XBRK</td>\n","      <td>SM-6XBRJ</td>\n","      <td>No</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>Train</td>\n","    </tr>\n","    <tr>\n","      <th>MSMB4LZ4</th>\n","      <td>M2071</td>\n","      <td>M2071C12</td>\n","      <td>metabolomics</td>\n","      <td>57.0</td>\n","      <td>2016-07-06</td>\n","      <td>13.0</td>\n","      <td>16</td>\n","      <td>ibdmdb</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>...</td>\n","      <td>No</td>\n","      <td>SM-B4LZ4</td>\n","      <td>SM-B4LZ5</td>\n","      <td>No</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>Test</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>5 rows Ã— 489 columns</p>\n","</div>"],"text/plain":["            Participant ID site_sub_coll     data_type  week_num  \\\n","External ID                                                        \n","CSM79HP4             C3022      C3022C11  metabolomics      21.0   \n","MSM9VZMM             M2084       M2084C1  metabolomics       0.0   \n","MSM79HDE             M2069       M2069C5  metabolomics       8.0   \n","PSM6XBRK             P6009       P6009C1  metabolomics       0.0   \n","MSMB4LZ4             M2071      M2071C12  metabolomics      57.0   \n","\n","            date_of_receipt  interval_days  visit_num Research Project  \\\n","External ID                                                              \n","CSM79HP4         2015-04-21           20.0         15           ibdmdb   \n","MSM9VZMM         2015-08-10            0.0          4           ibdmdb   \n","MSM79HDE         2015-06-02           14.0          8           ibdmdb   \n","PSM6XBRK         2014-11-12            0.0          4           ibdmdb   \n","MSMB4LZ4         2016-07-06           13.0         16           ibdmdb   \n","\n","             PDO Number  GSSR IDs  ...  Proteomics received at LBNL:  \\\n","External ID                        ...                                 \n","CSM79HP4            NaN       NaN  ...                            No   \n","MSM9VZMM            NaN       NaN  ...                            No   \n","MSM79HDE            NaN       NaN  ...                            No   \n","PSM6XBRK            NaN       NaN  ...                            No   \n","MSMB4LZ4            NaN       NaN  ...                            No   \n","\n","             Stool Sample ID:  Tube A  (EtOH)  \\\n","External ID                                     \n","CSM79HP4                             SM-79HP4   \n","MSM9VZMM                             SM-9VZMM   \n","MSM79HDE                             SM-79HDE   \n","PSM6XBRK                             SM-6XBRK   \n","MSMB4LZ4                             SM-B4LZ4   \n","\n","             Sample ID: Tube B (No preservative)  \\\n","External ID                                        \n","CSM79HP4                                SM-79HP5   \n","MSM9VZMM                                SM-9VZMN   \n","MSM79HDE                                SM-79HDF   \n","PSM6XBRK                                SM-6XBRJ   \n","MSMB4LZ4                                SM-B4LZ5   \n","\n","             Tube A and B received at Broad:  stool_id  smoking status  \\\n","External ID                                                              \n","CSM79HP4                                  No       NaN             NaN   \n","MSM9VZMM                                  No       NaN             NaN   \n","MSM79HDE                                  No       NaN             NaN   \n","PSM6XBRK                                  No       NaN             NaN   \n","MSMB4LZ4                                  No       NaN             NaN   \n","\n","             Number years smoked  Age when started smoking  \\\n","External ID                                                  \n","CSM79HP4                     NaN                       NaN   \n","MSM9VZMM                     NaN                       NaN   \n","MSM79HDE                     NaN                       NaN   \n","PSM6XBRK                     NaN                       NaN   \n","MSMB4LZ4                     NaN                       NaN   \n","\n","             How many cigarettes/cigars/etc. do you smoke per day?  Testing  \n","External ID                                                                  \n","CSM79HP4                                                   NaN        Train  \n","MSM9VZMM                                                   NaN        Train  \n","MSM79HDE                                                   NaN        Train  \n","PSM6XBRK                                                   NaN        Train  \n","MSMB4LZ4                                                   NaN         Test  \n","\n","[5 rows x 489 columns]"]},"execution_count":14,"metadata":{},"output_type":"execute_result"}],"source":"ihmp_data_path = '../../data/simulations/ihmp/'\nihmp_metadata = pd.read_csv('{}/sample-metadata.txt'.format(ihmp_data_path), sep='\\t', index_col=0)\nihmp_metadata.head(5)"},{"cell_type":"code","execution_count":15,"metadata":{},"outputs":[{"data":{"text/plain":["{'virome_virmap_analysis': 260 x 645 <class 'biom.table.Table'> with 2490 nonzero entries (1% dense),\n"," 'meta_t_ecs': 8291 x 805 <class 'biom.table.Table'> with 368525 nonzero entries (5% dense),\n"," 'meta_g_taxonomic_profiles': 1471 x 1638 <class 'biom.table.Table'> with 290658 nonzero entries (12% dense),\n"," 'shared_meta_g_taxonomic_profiles': 181 x 1630 <class 'biom.table.Table'> with 51942 nonzero entries (17% dense),\n"," 'HMP2_proteomics_ecs': 830 x 450 <class 'biom.table.Table'> with 56580 nonzero entries (15% dense),\n"," 'HMP2_metabolomics': 81867 x 546 <class 'biom.table.Table'> with 27348904 nonzero entries (61% dense)}"]},"execution_count":15,"metadata":{},"output_type":"execute_result"}],"source":"# import and match the data the same as in ../../simulations-benchmarking/3.0-ihmp-benchmarks.ipynb\n# we will also save each table with all the data for that table\n# for exploration non-jointly based on the joint samples\nihmp_tables = {omics_.split('/')[-1].split('.')[0]:rpca_table_processing(load_table(omics_),\n                                                                         min_sample_count=0,\n                                                                         min_feature_count=0,\n                                                                         min_feature_frequency=0)\n                for omics_ in glob.glob('{}/*.biom'.format(ihmp_data_path))\n                if not any ('_' + str(d_) in omics_ for d_ in [10, 8, 6, 4, 2, 1])}\n\n# add pathways instead of ECs (easier to interpret)\ntbl_ = pd.read_csv('{}/additional-data/pathabundances_3.tsv'.format(ihmp_data_path), sep='\\t', index_col=0)\ntbl_ = tbl_[[('UNINTEGRATED' not in x) and ('UNMAPPED' not in x) for x in tbl_.index]]\ntbl_.columns = [c.replace('_pathabundance_cpm','') for c in tbl_.columns]\ntbl_ = tbl_.loc[tbl_.sum(1) > 0, tbl_.sum(0) > 0]\ntbl_ = Table(tbl_.values, tbl_.index, tbl_.columns)\nihmp_tables['meta_t_ecs'] = tbl_.copy()\nihmp_tables"},{"cell_type":"code","execution_count":16,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["virome_virmap_analysis\n","meta_t_ecs\n","meta_g_taxonomic_profiles\n","shared_meta_g_taxonomic_profiles\n","HMP2_proteomics_ecs\n","HMP2_metabolomics\n"]},{"data":{"text/plain":["dict_keys(['virome_virmap_analysis', 'meta_t_ecs', 'meta_g_taxonomic_profiles', 'shared_meta_g_taxonomic_profiles', 'HMP2_proteomics_ecs', 'HMP2_metabolomics'])"]},"execution_count":16,"metadata":{},"output_type":"execute_result"}],"source":"# make table/metadata pair for each dataset of all samples in the data\ntables_metdata_unshared = {}\nfor omics_, table_ in ihmp_tables.items():\n    print(omics_)\n    metadata_omic = ihmp_metadata.copy()\n    table_omic = table_.copy()\n    shared_samps = set(table_omic.ids()) & set(ihmp_metadata.index)\n    table_omic = table_omic.filter(shared_samps)\n    table_omic = rpca_table_processing(table_omic, min_sample_count=0,\n                                       min_feature_count=0,\n                                       min_feature_frequency=0)\n    metadata_omic = metadata_omic.reindex(shared_samps)\n    table_omic_df = pd.DataFrame(table_omic.matrix_data.toarray(),\n                                 table_omic.ids('observation'),\n                                 table_omic.ids())\n    # these omics output in % abundance so have to use relative counts.\n    # possibly try to re-scale back to counts?\n    if omics_ in ['shared_meta_g_taxonomic_profiles','meta_t_ecs','HMP2_metabolomics']:\n        table_omic_df = table_omic_df.apply(closure)\n    tables_metdata_unshared[omics_] = [table_omic, table_omic_df, metadata_omic]\ntables_metdata_unshared.keys()"},{"cell_type":"code","execution_count":17,"metadata":{},"outputs":[{"data":{"text/plain":["CD        61\n","nonIBD    44\n","UC        30\n","Name: diagnosis, dtype: int64"]},"execution_count":17,"metadata":{},"output_type":"execute_result"}],"source":"shared_samps = set.intersection(*[set(t_.ids()) for t_ in ihmp_tables.values()]) & set(ihmp_metadata.index)\nmetadata_shared = ihmp_metadata.reindex(shared_samps)\nmetadata_shared['diagnosis_binned'] = [x.replace('UC','IBD').replace('CD','IBD')\n                                       for x in metadata_shared.diagnosis]\n# re-close metaG/T data\nmetadata_shared = metadata_shared.reindex(shared_samps)\ntables_shared = {t_k:t_.copy().filter(shared_samps) for t_k, t_ in ihmp_tables.items()}\nfor t_ in ['shared_meta_g_taxonomic_profiles','meta_t_ecs','HMP2_metabolomics']:\n    tbl_tmp = tables_shared[t_].to_dataframe().copy()\n    tbl_tmp = tbl_tmp.apply(closure)\n    tables_shared[t_] = Table(tbl_tmp.values, tbl_tmp.index, tbl_tmp.columns)\n\ntrain_, test_ = train_test_split(metadata_shared, shuffle=True,\n                                 stratify=metadata_shared['diagnosis'],\n                                 test_size=0.25)\nmetadata_shared['train_test'] = 'train'\nmetadata_shared.loc[test_.index, 'train_test'] = 'test'\n#metadata_shared.to_csv('../data/ihmp-sample-metadata-plus-train-tests.csv')\nmetadata_shared.diagnosis.value_counts()"},{"cell_type":"code","execution_count":18,"metadata":{},"outputs":[],"source":"# #save all tables in biom format\n# for k_, t_ in tables_shared.items():\n#    with open('{}iHMP/{}.biom'.format(save_path,k_), 'w') as f:\n#        f.write(t_.to_json(generated_by='BIOM-Format'))"},{"cell_type":"markdown","metadata":{},"source":"### UC cohort two"},{"cell_type":"code","execution_count":20,"metadata":{},"outputs":[],"source":"uc_data_path = '../../data/case-studies/uc-severity-multiomics/Cohort_two/'\nuc_metadata = pd.read_csv('{}metadata_cohort_two_revised.txt'.format(uc_data_path), sep='\\t', index_col=0)"},{"cell_type":"code","execution_count":4,"metadata":{},"outputs":[{"data":{"text/plain":["{'metaproteomics': 108080 x 174 <class 'biom.table.Table'> with 5091405 nonzero entries (27% dense),\n"," 'metagenomics': 3568 x 174 <class 'biom.table.Table'> with 163422 nonzero entries (26% dense),\n"," 'metabolomics': 1928 x 174 <class 'biom.table.Table'> with 58524 nonzero entries (17% dense)}"]},"execution_count":4,"metadata":{},"output_type":"execute_result"}],"source":"uc_tables = {'metaproteomics': load_table('{}metaproteomics_cohort_two_matched.biom'.format(uc_data_path)),\n             'metagenomics': load_table('{}metagenomics_cohort_two_matched.biom'.format(uc_data_path)),\n             'metabolomics': load_table('{}metabolomics_cohort_two_matched.biom'.format(uc_data_path))}\nuc_tables"},{"cell_type":"code","execution_count":5,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["metaproteomics\n","metagenomics\n","metabolomics\n"]},{"data":{"text/plain":["dict_keys(['metaproteomics', 'metagenomics', 'metabolomics'])"]},"execution_count":5,"metadata":{},"output_type":"execute_result"}],"source":"# make table/metadata pair for each dataset of all samples in the data\nuc_tables_metdata_unshared = {}\nfor omics_, table_ in uc_tables.items():\n    print(omics_)\n    metadata_omic = uc_metadata.copy()\n    table_omic = table_.copy()\n    shared_samps = set(table_omic.ids()) & set(uc_metadata.index)\n    table_omic = table_omic.filter(shared_samps)\n    table_omic = rpca_table_processing(table_omic, min_sample_count=0,\n                                       min_feature_count=0,\n                                       min_feature_frequency=0)\n    metadata_omic = metadata_omic.reindex(shared_samps)\n    table_omic_df = pd.DataFrame(table_omic.matrix_data.toarray(),\n                                 table_omic.ids('observation'),\n                                 table_omic.ids())\n    # these omics output in % abundance so have to use relative counts.\n    # possibly try to re-scale back to counts?\n    if omics_ in ['metaproteomics','metagenomics','metabolomics']:\n        table_omic_df = table_omic_df.apply(closure)\n    uc_tables_metdata_unshared[omics_] = [table_omic, table_omic_df, metadata_omic]\nuc_tables_metdata_unshared.keys()"},{"cell_type":"code","execution_count":6,"metadata":{},"outputs":[{"data":{"text/plain":["Diagnosis\n","IBD                157\n","Healthy_control     16\n","Name: count, dtype: int64"]},"execution_count":6,"metadata":{},"output_type":"execute_result"}],"source":"shared_samps = list(set.intersection(*[set(t_.ids()) for t_ in uc_tables.values()]) & set(uc_metadata.index))\nuc_metadata_shared = uc_metadata.reindex(shared_samps)\nuc_metadata_shared['Diagnosis'] = [x.replace('UC','IBD').replace('CD','IBD')\n                                for x in uc_metadata_shared.Diagnosis]\nuc_metadata_shared = uc_metadata_shared.reindex(shared_samps)\ntables_shared = {t_k:t_.copy().filter(shared_samps) for t_k, t_ in uc_tables.items()}\n\ntrain_, test_ = train_test_split(uc_metadata_shared, shuffle=True,\n                                 stratify=uc_metadata_shared['Diagnosis'],\n                                 test_size=0.25)\nuc_metadata_shared['train_test'] = 'train'\nuc_metadata_shared.loc[test_.index, 'train_test'] = 'test'\nuc_metadata_shared.to_csv('../../data/case-studies/time-ihmp-uc/uc-metadata-plus-train-tests.csv')\nuc_metadata_shared.Diagnosis.value_counts()"},{"cell_type":"code","execution_count":7,"metadata":{},"outputs":[],"source":"#save all tables in biom format\n# for k_, t_ in tables_shared.items():\n#    with open('{}UC/{}.biom'.format(save_path,k_), 'w') as f:\n#        f.write(t_.to_json(generated_by='BIOM-Format'))"}],"metadata":{"kernelspec":{"display_name":"Python 3 (ipykernel)","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.9.19"}},"nbformat":4,"nbformat_minor":4}