{"cells":[{"cell_type":"code","execution_count":1,"metadata":{},"outputs":[],"source":"import glob\nimport pandas as pd\nimport qiime2 as q2\nimport time\nfrom qiime2.plugins.mmvec.actions import paired_omics                                   \nfrom biom import load_table\nimport json\nimport itertools\n\n#turn off warnings\nimport warnings\nwarnings.filterwarnings(\"ignore\")\n#turn off tensorflow warnings\nimport os\nos.environ['TF_CPP_MIN_LOG_LEVEL'] = '2' "},{"cell_type":"code","execution_count":2,"metadata":{},"outputs":[],"source":"##calculate # of features needed to match percentages\ndef get_n_features(p_features, total_features, subtotal_features):\n    total_fts = [int(0.01 * p * total_features) for p in p_features]\n    prot_fts = [fts - subtotal_features for fts in total_fts]\n    return total_fts, prot_fts"},{"cell_type":"code","execution_count":3,"metadata":{},"outputs":[{"data":{"text/plain":["{'metagenomics': 3568 x 173 <class 'biom.table.Table'> with 162126 nonzero entries (26% dense),\n"," 'metabolomics': 1928 x 173 <class 'biom.table.Table'> with 58132 nonzero entries (17% dense),\n"," 'metaproteomics': 108080 x 173 <class 'biom.table.Table'> with 5065165 nonzero entries (27% dense)}"]},"execution_count":3,"metadata":{},"output_type":"execute_result"}],"source":"#load the data\ndata_path = '../../data/simulations/ihmp/'\nuc_omic_keys = [omics_.split('/')[-1].split('.')[0] for omics_ in glob.glob('{}UC/*.biom'.format(data_path))]\n\nuc_metadata_shared = pd.read_csv('../../data/case-studies/time-ihmp-uc/uc-metadata-plus-train-tests.csv', index_col=0)\nuc_tables_shared = {k_: load_table('{}UC/{}.biom'.format(data_path, k_)) for k_ in uc_omic_keys}\nuc_tables_shared"},{"cell_type":"code","execution_count":4,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["113576 5496\n"]}],"source":"#add number of features across tables\nuc_n_feat = 0\nfor k, v in uc_tables_shared.items():\n    uc_n_feat += v.shape[0]\n\nuc_subtotal = uc_n_feat - uc_tables_shared['metaproteomics'].shape[0]\nprint(uc_n_feat, uc_subtotal)"},{"cell_type":"code","execution_count":5,"metadata":{},"outputs":[{"data":{"text/plain":["array([10.46, 11.35, 13.14, 14.94, 19.41, 23.89])"]},"execution_count":5,"metadata":{},"output_type":"execute_result"}],"source":"#load mmvec (iHMP) results to get % features\nmmvec_ihmp = pd.read_csv('../../data/case-studies/time-ihmp-uc/ihmp-runtime-all.csv', index_col=0)\nmmvec_ihmp = mmvec_ihmp[mmvec_ihmp.method == 'MMvec']\nperct = mmvec_ihmp['% features'].values\nperct"},{"cell_type":"code","execution_count":6,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["[11880, 12890, 14923, 16968, 22045, 27133]\n","[6384, 7394, 9427, 11472, 16549, 21637]\n"]}],"source":"#print the number of proteomic features needed to match the percentages\n#as well as the total # of features after adding the subset of proteomics\nall_factors = list(perct)\nuc_total_n, uc_prot_n = get_n_features(all_factors, uc_n_feat, uc_subtotal)\nprint(uc_total_n)\nprint(uc_prot_n)"},{"cell_type":"code","execution_count":7,"metadata":{},"outputs":[],"source":"##formatting for running MMvec\n#change index name to 'sample id'\nuc_metadata_shared.index.name = 'sample id'\n\n#change train to Train and test to Test\nuc_metadata_shared['train_test_mmvec'] = uc_metadata_shared['train_test'].apply(lambda x: 'Train' if x == 'train' else 'Test')"},{"cell_type":"code","execution_count":8,"metadata":{},"outputs":[{"data":{"text/plain":["[('metagenomics', 'metagenomics'),\n"," ('metabolomics', 'metabolomics'),\n"," ('metaproteomics', 'metaproteomics'),\n"," ('metagenomics', 'metabolomics'),\n"," ('metagenomics', 'metaproteomics'),\n"," ('metabolomics', 'metagenomics'),\n"," ('metabolomics', 'metaproteomics'),\n"," ('metaproteomics', 'metagenomics'),\n"," ('metaproteomics', 'metabolomics')]"]},"execution_count":8,"metadata":{},"output_type":"execute_result"}],"source":"#create list with all possible pairs of omics\nomic_keys = list(uc_tables_shared.keys())\npairs = [('metagenomics', 'metagenomics'), \n         ('metabolomics', 'metabolomics'),\n         ('metaproteomics', 'metaproteomics')] + list(itertools.permutations(omic_keys, 2))\npairs"},{"cell_type":"code","execution_count":9,"metadata":{},"outputs":[],"source":"#create list with and without proteomics\npairs_no_prot = [pair for pair in pairs if 'metaproteomics' not in pair]\npairs_prot = [pair for pair in pairs if 'metaproteomics' in pair]"},{"cell_type":"code","execution_count":10,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["Proteomics: 1.0% of features: 1080\n","Proteomics: 3.0% of features: 3242\n","Proteomics: 4.0% of features: 4323\n","Proteomics: 5.0% of features: 5404\n","Proteomics: 10.0% of features: 10808\n"]}],"source":"#generate subsets of metabolomics data\nproteomics_all = uc_tables_shared['metaproteomics'].copy()\n\nproteomics_sub = {}\nfor factor in [0.01, 0.03, 0.04, 0.05, 0.1]:\n    #get metabolomics table\n    prot = uc_tables_shared['metaproteomics'].copy()\n    #number of features to keep\n    n_features = int(prot.shape[0]*factor)\n    #take first n_features\n    feature_ids = list(prot.ids(axis='observation'))[:n_features]\n    #filter table\n    proteomics_sub[factor] = prot.filter(feature_ids, axis='observation')\n    #sanity check\n    print('Proteomics: {}% of features: {}'.format(factor*100, proteomics_sub[factor].shape[0]))"},{"cell_type":"code","execution_count":11,"metadata":{},"outputs":[],"source":"uc_metadata_shared_short = uc_metadata_shared.copy()\nuc_metadata_shared_short = uc_metadata_shared_short[['train_test_mmvec']]"},{"cell_type":"code","execution_count":20,"metadata":{},"outputs":[],"source":"# runtime = {}\n# runtime_total = {}"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"# #run mmvec for each pair of omics without proteomics\n# factor = 0\n# time_factor = 0\n# for pair in pairs_no_prot:  \n#     print(pair)\n#     #get tables and set to correct format\n#     t1 = uc_tables_shared[pair[0]]\n#     t2 = uc_tables_shared[pair[1]]\n#     t1_q2 = q2.Artifact.import_data('FeatureTable[Frequency]', t1)\n#     t2_q2 = q2.Artifact.import_data('FeatureTable[Frequency]', t2)\n\n#     #run and time mmvec\n#     time_start = time.perf_counter()\n#     mmvec_res = paired_omics(t1_q2, t2_q2, \n#                             summary_interval=1,\n#                             metadata=q2.Metadata(uc_metadata_shared_short),\n#                             training_column='train_test_mmvec',\n#                             min_feature_count=10)\n#     time_elapsed = (time.perf_counter() - time_start)\n#     time_factor += time_elapsed\n#     print('Time elapsed: ', round(time_elapsed,4))\n#     print()\n#     #save output\n#     runtime[\"{}-{}\".format(pair[0], pair[1])] = time_elapsed\n#     with open(\"../results/mmvec-runtime-uc.json\", \"w\") as outfile: \n#         json.dump(runtime, outfile)\n\n# runtime_total[factor] = time_factor\n# print('Total time elapsed: ', round(time_factor,4))\n\n# #save mmvec results as json/pickle\n# with open(\"../results/mmvec-runtime-total-uc.json\", \"w\") as outfile: \n#    json.dump(runtime_total, outfile)"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"# #sanity check\n# print(runtime)\n# print(runtime_total)"},{"cell_type":"code","execution_count":12,"metadata":{},"outputs":[],"source":"#load mmvec results\nwith open(\"../results/mmvec-runtime-total-uc.json\", \"r\") as infile:\n    runtime_total = json.load(infile)\n\nwith open(\"../results/mmvec-runtime-uc.json\", \"r\") as infile:\n    runtime = json.load(infile)"},{"cell_type":"code","execution_count":13,"metadata":{},"outputs":[{"data":{"text/plain":["{'metagenomics-metagenomics': 1563.6010937170067,\n"," 'metabolomics-metabolomics': 179.47512698400533,\n"," 'metagenomics-metabolomics': 957.9216301410052,\n"," 'metabolomics-metagenomics': 334.7416218419967,\n"," 'metaproteomics0.01-metaproteomics0.01': 232.43496004698682,\n"," 'metagenomics-metaproteomics0.01': 499.39272144700226,\n"," 'metabolomics-metaproteomics0.01': 105.30647485500958,\n"," 'metaproteomics0.01-metagenomics': 582.3444733399956,\n"," 'metaproteomics0.01-metabolomics': 372.5092970100086,\n"," 'metaproteomics0.03-metaproteomics0.03': 3680.409118486001,\n"," 'metagenomics-metaproteomics0.03': 1343.1144764290075,\n"," 'metabolomics-metaproteomics0.03': 252.66649123700336,\n"," 'metaproteomics0.03-metagenomics': 3071.323532092996,\n"," 'metaproteomics0.03-metabolomics': 1140.475918815995,\n"," 'metagenomics-metaproteomics0.05': 5627.837036309997,\n"," 'metabolomics-metaproteomics0.05': 457.6624763409782,\n"," 'metaproteomics0.05-metabolomics': 3228.7686570439837,\n"," 'metaproteomics0.05-metaproteomics0.05': 360000,\n"," 'metagenomics-metaproteomics0.04': 3564.4600208510237,\n"," 'metabolomics-metaproteomics0.04': 420.6446165340021,\n"," 'metaproteomics0.04-metagenomics': 19972.53910159599,\n"," 'metaproteomics0.04-metabolomics': 1800.3676350360038,\n"," 'metagenomics-metaproteomics0.06_only': 26016.035002462995,\n"," 'metabolomics-metaproteomics0.06_only': 481.7244231939985,\n"," 'metaproteomics0.06-metabolomics_only': 3592.27986486601}"]},"execution_count":13,"metadata":{},"output_type":"execute_result"}],"source":"runtime"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"for factor in [0.01, 0.03]: #0.04 onwards is too slow\n    print('\\n### Proteomic Features: {}% ###'.format(factor*100))\n    prot_table = proteomics_sub[factor]\n    table_to_use = uc_tables_shared.copy()\n    table_to_use['metaproteomics'] = prot_table\n    print('Proteomics n features: {}'.format(table_to_use['metaproteomics'].shape[0]))\n\n    time_factor = 0\n    for pair in pairs_prot:\n        print(pair)\n        \n        #get tables and set to correct format\n        t1 = table_to_use[pair[0]]\n        t2 = table_to_use[pair[1]]\n        t1_q2 = q2.Artifact.import_data('FeatureTable[Frequency]', t1)\n        t2_q2 = q2.Artifact.import_data('FeatureTable[Frequency]', t2)\n\n        #run and time mmvec\n        time_start = time.perf_counter()\n        mmvec_res = paired_omics(t1_q2, t2_q2, \n                                summary_interval=1,\n                                metadata=q2.Metadata(uc_metadata_shared_short),\n                                training_column='train_test_mmvec',\n                                min_feature_count=10)\n        time_elapsed = (time.perf_counter() - time_start)\n        time_factor += time_elapsed\n        print('Time elapsed: ', round(time_elapsed,2))\n        print()\n\n        # update pair name before saving\n        if pair[0] == 'metaproteomics':\n            pair = ('metaproteomics{}'.format(factor), pair[1])\n        if pair[1] == 'metaproteomics':\n            pair = (pair[0], 'metaproteomics{}'.format(factor))\n        #save output\n        runtime[\"{}-{}\".format(pair[0], pair[1])] = time_elapsed\n        with open(\"../../results/case-studies/time-ihmp-uc/mmvec-runtime-uc.json\", \"w\") as outfile: \n            json.dump(runtime, outfile)\n    \n    runtime_total[factor] = time_factor\n    print('Total time elapsed: {} secs ({} mins)'.format(round(time_factor, 2),\n                                                         round(time_factor/60, 2)))\n    \n    #save mmvec results as json/pickle\n    with open(\"../../data/case-studies/time-ihmp-uc/mmvec-runtime-total-uc.json\", \"w\") as outfile: \n        json.dump(runtime_total, outfile)\n    "},{"cell_type":"code","execution_count":14,"metadata":{},"outputs":[],"source":"pairs_prot_sub = [#('metaproteomics', 'metaproteomics'),\n                    ('metagenomics', 'metaproteomics'),\n                    ('metabolomics', 'metaproteomics'),\n                    ('metaproteomics', 'metagenomics'),\n                    ('metaproteomics', 'metabolomics')]\n\nfor factor in [0.04]:\n    print('\\n### Proteomic Features: {}% ###'.format(factor*100))\n    prot_table = proteomics_sub[factor]\n    table_to_use = uc_tables_shared.copy()\n    table_to_use['metaproteomics'] = prot_table\n    print('Proteomics n features: {}'.format(table_to_use['metaproteomics'].shape[0]))\n\n    time_factor = 0\n    for pair in pairs_prot_sub:\n        print(pair)\n        \n        #get tables and set to correct format\n        t1 = table_to_use[pair[0]]\n        t2 = table_to_use[pair[1]]\n        t1_q2 = q2.Artifact.import_data('FeatureTable[Frequency]', t1)\n        t2_q2 = q2.Artifact.import_data('FeatureTable[Frequency]', t2)\n\n        #run and time mmvec\n        time_start = time.perf_counter()\n        mmvec_res = paired_omics(t1_q2, t2_q2, \n                                summary_interval=1,\n                                metadata=q2.Metadata(uc_metadata_shared_short),\n                                training_column='train_test_mmvec',\n                                min_feature_count=10)\n        time_elapsed = (time.perf_counter() - time_start)\n        time_factor += time_elapsed\n        print('Time elapsed: ', round(time_elapsed,2))\n        print()\n\n        # update pair name before saving\n        if pair[0] == 'metaproteomics':\n            pair = ('metaproteomics{}'.format(factor), pair[1])\n        if pair[1] == 'metaproteomics':\n            pair = (pair[0], 'metaproteomics{}'.format(factor))\n        #save output\n        runtime[\"{}-{}\".format(pair[0], pair[1])] = time_elapsed\n        with open(\"../../data/case-studies/time-ihmp-uc/mmvec-runtime-uc.json\", \"w\") as outfile: \n            json.dump(runtime, outfile)\n    \n    runtime_total[factor] = time_factor\n    print('Total time elapsed: {} secs ({} mins)'.format(round(time_factor, 2),\n                                                         round(time_factor/60, 2)))\n    \n    #save mmvec results as json/pickle\n    with open(\"../../data/case-studies/time-ihmp-uc/mmvec-runtime-total-uc.json\", \"w\") as outfile: \n        json.dump(runtime_total, outfile)\n    "},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"pairs_prot_sub = [#('metaproteomics', 'metaproteomics'),\n                    ('metagenomics', 'metaproteomics'),\n                    ('metabolomics', 'metaproteomics'),\n                    #('metaproteomics', 'metagenomics'),\n                    ('metaproteomics', 'metabolomics')]\n\nfor factor in [0.05]:\n    print('\\n### Proteomic Features: {}% ###'.format(factor*100))\n    prot_table = proteomics_sub[factor]\n    table_to_use = uc_tables_shared.copy()\n    table_to_use['metaproteomics'] = prot_table\n    print('Proteomics n features: {}'.format(table_to_use['metaproteomics'].shape[0]))\n\n    time_factor = 0\n    for pair in pairs_prot_sub:\n        print(pair)\n        \n        #get tables and set to correct format\n        t1 = table_to_use[pair[0]]\n        t2 = table_to_use[pair[1]]\n        t1_q2 = q2.Artifact.import_data('FeatureTable[Frequency]', t1)\n        t2_q2 = q2.Artifact.import_data('FeatureTable[Frequency]', t2)\n\n        #run and time mmvec\n        time_start = time.perf_counter()\n        mmvec_res = paired_omics(t1_q2, t2_q2, \n                                summary_interval=1,\n                                metadata=q2.Metadata(uc_metadata_shared_short),\n                                training_column='train_test_mmvec',\n                                min_feature_count=10)\n        time_elapsed = (time.perf_counter() - time_start)\n        time_factor += time_elapsed\n        print('Time elapsed: ', round(time_elapsed,2))\n        print()\n\n        # update pair name before saving\n        if pair[0] == 'metaproteomics':\n            pair = ('metaproteomics{}'.format(factor), pair[1])\n        if pair[1] == 'metaproteomics':\n            pair = (pair[0], 'metaproteomics{}'.format(factor))\n        #save output\n        runtime[\"{}-{}\".format(pair[0], pair[1])] = time_elapsed\n        with open(\"../../data/case-studies/time-ihmp-uc/mmvec-runtime-uc.json\", \"w\") as outfile: \n            json.dump(runtime, outfile)\n    \n    runtime_total[factor] = time_factor\n    print('Total time elapsed: {} secs ({} mins)'.format(round(time_factor, 2),\n                                                         round(time_factor/60, 2)))\n    \n    #save mmvec results as json/pickle\n    with open(\"../../data/case-studies/time-ihmp-uc/mmvec-runtime-total-uc.json\", \"w\") as outfile: \n        json.dump(runtime_total, outfile)\n    "},{"cell_type":"markdown","metadata":{},"source":"### Add runtimes"},{"cell_type":"code","execution_count":16,"metadata":{},"outputs":[],"source":"#load mmvec results\nwith open(\"../../data/case-studies/time-ihmp-uc/mmvec-runtime-total-uc.json\", \"r\") as infile:\n    runtime_total = json.load(infile)\n\nwith open(\"../../data/case-studies/time-ihmp-uc/mmvec-runtime-uc.json\", \"r\") as infile:\n    runtime = json.load(infile)"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"## add an estimated time for runtime_total with factor = 0.5 for ('metaproteomics', 'metaproteomics')\n#100hrs = 100*60*60 = 360000\n#runtime['metaproteomics0.05-metaproteomics0.05'] = 360000\n#runtime"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"#now, update runtime_total to include a new pair\nruntime_prot_004 = {k: v for k, v in runtime.items() if 'metaproteomics0.04' in k}\nsum(runtime_prot_004.values())\n#need to add runtime_total[0] to this sum \nruntime_total['0.04'] += runtime_total['0']\nruntime_total\n#add to total runtime\n#runtime_total['0.05'] = sum(runtime_prot_004.values())"},{"cell_type":"code","execution_count":26,"metadata":{},"outputs":[],"source":"#now, update runtime_total to include another new pair\nruntime_prot = {k: v for k, v in runtime.items() if 'metaproteomics0.05' in k}\n#add to total runtime\nruntime_total['0.05'] = sum(runtime_prot.values())"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"#need to add runtime_total[0] to each individual runtime \n#runtime_total[0] is the time that all pairs excluding metaproteomics took to run\nprint('runtime_total[0]: ', runtime_total['0'])\nprint()\nfor key in ['0.05']:\n    print('runtime_total[{}]: '.format(key), runtime_total[key])\n    runtime_total[key] += runtime_total['0']\n    print('Updated runtime_total[{}]: '.format(key), runtime_total[key])"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"#need to add runtime_total[0] to each individual runtime \n#runtime_total[0] is the time that all pairs excluding metaproteomics took to run\nprint('runtime_total[0]: ', runtime_total['0'])\nprint()\nfor key in ['0.01', '0.03']: #ignore 0.05 output\n    print('runtime_total[{}]: '.format(key), runtime_total[key])\n    runtime_total[key] += runtime_total['0']\n    print('Updated runtime_total[{}]: '.format(key), runtime_total[key])"},{"cell_type":"code","execution_count":19,"metadata":{},"outputs":[],"source":"# save final set of mmvec results\nwith open(\"../../data/case-studies/time-ihmp-uc/mmvec-runtime-total-uc.json\", \"w\") as outfile: \n   json.dump(runtime_total, outfile)\n\nwith open(\"../../data/case-studies/time-ihmp-uc/mmvec-runtime-uc.json\", \"w\") as outfile: \n  json.dump(runtime, outfile)"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":"#print time in minutes\nruntime_total_minutes = {k: v/60 for k, v in runtime_total.items()}\nruntime_total_minutes\n#runtime_minutes = {k: v/60 for k, v in runtime.items()}"}],"metadata":{"kernelspec":{"display_name":"Python 3 (ipykernel)","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.9.19"}},"nbformat":4,"nbformat_minor":4}