{"cells":[{"cell_type":"code","execution_count":1,"metadata":{},"outputs":[],"source":"import pandas as pd\nfrom gemelli.preprocessing import rclr_transformation\nfrom scipy.spatial import distance\nfrom scipy.linalg import svd\nfrom biom import Table\nfrom scipy.sparse.linalg import svds\nfrom skbio import OrdinationResults, DistanceMatrix\nimport qiime2 as q2\nfrom qiime2.plugins.emperor.actions import biplot\nfrom biom.util import biom_open\nimport numpy as np\n\nimport matplotlib.pyplot as plt\nimport seaborn as sns\n%matplotlib inline"},{"cell_type":"code","execution_count":2,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["number shared samples: 101\n"]},{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>diet</th>\n","      <th>digestive_strategy</th>\n","      <th>HostSpecies</th>\n","      <th>HostSubSpecies</th>\n","      <th>KT_lineage</th>\n","      <th>SpeciesCategories</th>\n","      <th>Taxonomy</th>\n","      <th>diet_pri</th>\n","      <th>Collected_by</th>\n","      <th>Day</th>\n","      <th>...</th>\n","      <th>File_lipids_pos</th>\n","      <th>Lipids_order</th>\n","      <th>File_VFA</th>\n","      <th>VFA_order</th>\n","      <th>File_GCMS</th>\n","      <th>GCMS_order</th>\n","      <th>File_polar.1</th>\n","      <th>Sex</th>\n","      <th>collection_time</th>\n","      <th>lineage_and_digestive_strategy</th>\n","    </tr>\n","    <tr>\n","      <th>sample_name</th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>Chimpanzee_11</th>\n","      <td>Omnivores</td>\n","      <td>Simplegut</td>\n","      <td>Chimp</td>\n","      <td>Chimp</td>\n","      <td>Primates</td>\n","      <td>Chimps</td>\n","      <td>Pan_troglodytes</td>\n","      <td>Primates</td>\n","      <td>Ruthie</td>\n","      <td>18.0</td>\n","      <td>...</td>\n","      <td>lipids-230718-pos-69.mzXML</td>\n","      <td>103.0</td>\n","      <td>12_26_2018-inj-101</td>\n","      <td>101.0</td>\n","      <td>2019-01-21-inj-142.mzML</td>\n","      <td>142.0</td>\n","      <td>Sample-X-76</td>\n","      <td>F</td>\n","      <td>0-2hr</td>\n","      <td>Primates-Simplegut</td>\n","    </tr>\n","    <tr>\n","      <th>Rhino_6</th>\n","      <td>Herbivores</td>\n","      <td>Hindgut</td>\n","      <td>Rhino</td>\n","      <td>Rhino</td>\n","      <td>Perissodactyla</td>\n","      <td>Rhinos</td>\n","      <td>Ceratotherium_simum</td>\n","      <td>Herbivores</td>\n","      <td>Stav</td>\n","      <td>6.0</td>\n","      <td>...</td>\n","      <td>lipids-230718-pos-77.mzXML</td>\n","      <td>39.0</td>\n","      <td>12_26_2018-inj-38</td>\n","      <td>38.0</td>\n","      <td>2019-01-28-inj-210.mzML</td>\n","      <td>210.0</td>\n","      <td>Sample-2A6</td>\n","      <td>M</td>\n","      <td>0-2hr</td>\n","      <td>Perissodactyla-Hindgut</td>\n","    </tr>\n","    <tr>\n","      <th>Gorilla_5</th>\n","      <td>Herbivores</td>\n","      <td>Simplegut</td>\n","      <td>Gorilla</td>\n","      <td>Gorilla</td>\n","      <td>Primates</td>\n","      <td>Gorillas</td>\n","      <td>Gorilla_gorilla</td>\n","      <td>Primates</td>\n","      <td>Ruthie</td>\n","      <td>25.0</td>\n","      <td>...</td>\n","      <td>lipids-230718-pos-52.mzXML</td>\n","      <td>119.0</td>\n","      <td>12_27_2018-inj-110</td>\n","      <td>110.0</td>\n","      <td>2019-01-16-inj-85.mzML</td>\n","      <td>85.0</td>\n","      <td>Sample-2B1</td>\n","      <td>F</td>\n","      <td>0-2hr</td>\n","      <td>Primates-Simplegut</td>\n","    </tr>\n","    <tr>\n","      <th>Sheep_1</th>\n","      <td>Herbivores</td>\n","      <td>Foregut</td>\n","      <td>Sheep</td>\n","      <td>Sheep</td>\n","      <td>Cetartiodactyla</td>\n","      <td>Ruminants</td>\n","      <td>Ovis_aries</td>\n","      <td>Herbivores</td>\n","      <td>Stav+Elie</td>\n","      <td>11.0</td>\n","      <td>...</td>\n","      <td>lipids-230718-pos-1.mzXML</td>\n","      <td>7.0</td>\n","      <td>12_25_2018-inj-6</td>\n","      <td>6.0</td>\n","      <td>2019-01-13-inj-6.mzML</td>\n","      <td>6.0</td>\n","      <td>Sample-1A4</td>\n","      <td>NaN</td>\n","      <td>0-2hr</td>\n","      <td>Cetartiodactyla-Foregut</td>\n","    </tr>\n","    <tr>\n","      <th>Rhino_10</th>\n","      <td>Herbivores</td>\n","      <td>Hindgut</td>\n","      <td>Rhino</td>\n","      <td>Rhino</td>\n","      <td>Perissodactyla</td>\n","      <td>Rhinos</td>\n","      <td>Ceratotherium_simum</td>\n","      <td>Herbivores</td>\n","      <td>Stav+Ishay</td>\n","      <td>13.0</td>\n","      <td>...</td>\n","      <td>lipids-230718-pos-81.mzXML</td>\n","      <td>14.0</td>\n","      <td>12_25_2018-inj-14</td>\n","      <td>14.0</td>\n","      <td>2019-01-13-inj-14.mzML</td>\n","      <td>14.0</td>\n","      <td>Sample-1E3</td>\n","      <td>F</td>\n","      <td>0-2hr</td>\n","      <td>Perissodactyla-Hindgut</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>5 rows Ã— 26 columns</p>\n","</div>"],"text/plain":["                     diet digestive_strategy HostSpecies HostSubSpecies  \\\n","sample_name                                                               \n","Chimpanzee_11   Omnivores          Simplegut       Chimp          Chimp   \n","Rhino_6        Herbivores            Hindgut       Rhino          Rhino   \n","Gorilla_5      Herbivores          Simplegut     Gorilla        Gorilla   \n","Sheep_1        Herbivores            Foregut       Sheep          Sheep   \n","Rhino_10       Herbivores            Hindgut       Rhino          Rhino   \n","\n","                    KT_lineage SpeciesCategories             Taxonomy  \\\n","sample_name                                                             \n","Chimpanzee_11         Primates            Chimps      Pan_troglodytes   \n","Rhino_6         Perissodactyla            Rhinos  Ceratotherium_simum   \n","Gorilla_5             Primates          Gorillas      Gorilla_gorilla   \n","Sheep_1        Cetartiodactyla         Ruminants           Ovis_aries   \n","Rhino_10        Perissodactyla            Rhinos  Ceratotherium_simum   \n","\n","                 diet_pri Collected_by   Day  ...             File_lipids_pos  \\\n","sample_name                                   ...                               \n","Chimpanzee_11    Primates       Ruthie  18.0  ...  lipids-230718-pos-69.mzXML   \n","Rhino_6        Herbivores         Stav   6.0  ...  lipids-230718-pos-77.mzXML   \n","Gorilla_5        Primates       Ruthie  25.0  ...  lipids-230718-pos-52.mzXML   \n","Sheep_1        Herbivores    Stav+Elie  11.0  ...   lipids-230718-pos-1.mzXML   \n","Rhino_10       Herbivores   Stav+Ishay  13.0  ...  lipids-230718-pos-81.mzXML   \n","\n","               Lipids_order            File_VFA VFA_order  \\\n","sample_name                                                 \n","Chimpanzee_11         103.0  12_26_2018-inj-101     101.0   \n","Rhino_6                39.0   12_26_2018-inj-38      38.0   \n","Gorilla_5             119.0  12_27_2018-inj-110     110.0   \n","Sheep_1                 7.0    12_25_2018-inj-6       6.0   \n","Rhino_10               14.0   12_25_2018-inj-14      14.0   \n","\n","                             File_GCMS GCMS_order File_polar.1  Sex  \\\n","sample_name                                                           \n","Chimpanzee_11  2019-01-21-inj-142.mzML      142.0  Sample-X-76    F   \n","Rhino_6        2019-01-28-inj-210.mzML      210.0   Sample-2A6    M   \n","Gorilla_5       2019-01-16-inj-85.mzML       85.0   Sample-2B1    F   \n","Sheep_1          2019-01-13-inj-6.mzML        6.0   Sample-1A4  NaN   \n","Rhino_10        2019-01-13-inj-14.mzML       14.0   Sample-1E3    F   \n","\n","              collection_time  lineage_and_digestive_strategy  \n","sample_name                                                    \n","Chimpanzee_11           0-2hr              Primates-Simplegut  \n","Rhino_6                 0-2hr          Perissodactyla-Hindgut  \n","Gorilla_5               0-2hr              Primates-Simplegut  \n","Sheep_1                 0-2hr         Cetartiodactyla-Foregut  \n","Rhino_10                0-2hr          Perissodactyla-Hindgut  \n","\n","[5 rows x 26 columns]"]},"execution_count":2,"metadata":{},"output_type":"execute_result"}],"source":"metadata = pd.read_csv('../../data/case-studies/mammalian/metadata.txt', index_col=0, sep='\\t')\nmetadata_noqc = metadata[~metadata.SpeciesCategories.isin(['QC','Blank'])]\nmicrobiome = pd.read_csv('../../data/case-studies/mammalian/16S-ASV-table.csv', index_col=0).T\nlipids_neg = pd.read_csv('../../data/case-studies/mammalian/2018-12-18-lipids-neg_quant.txt', index_col=0, sep='\\t').drop(['row m/z','row retention time'], axis=1)\nlipids_pos = pd.read_csv('../../data/case-studies/mammalian/2018-12-18-lipids-pos_quant.txt', index_col=0, sep='\\t').drop(['row m/z','row retention time'], axis=1)\nlipids_polar = pd.read_csv('../../data/case-studies/mammalian/2019-07-04-polar-filter-singletons_quant.txt', index_col=0, sep='\\t').drop(['row m/z','row retention time'], axis=1)\n# gcms metabolomics filter\ngcms_metabolomics = pd.read_csv('../../data/case-studies/mammalian/gcms-quantification.csv', index_col=0).drop(['row m/z','row retention time'], axis=1)\ngcms_bscore = pd.read_csv('../../data/case-studies/mammalian/gcms-clustersummary.tsv', index_col=0, sep='\\t')\ngcms_bscore_filtered = gcms_bscore.loc[gcms_bscore['Balance Score'].astype(float) >= 50, :]\ngcms_metabolomics = gcms_metabolomics.loc[gcms_bscore_filtered.index, :]\n\n\ndef inverse_mapping(f):\n    return f.__class__(map(reversed, f.items()))\n\ngcms_map = inverse_mapping(metadata['File_GCMS'].to_dict())\nlippos_map = inverse_mapping(metadata['File_lipids_pos'].to_dict())\nlipneg_map = inverse_mapping(metadata['File_lipids_neg'].to_dict())\npolar_map = inverse_mapping(metadata['File_polar.1'].to_dict())\n\ngcms_metabolomics.columns = [gcms_map[c.split(' ')[0]] for c in gcms_metabolomics.columns]\nlipids_pos.columns = [lippos_map[c.split(' ')[0]] for c in lipids_pos.columns]\nlipids_neg.columns = [lipneg_map[c.split(' ')[0]] for c in lipids_neg.columns]\nlipids_polar.columns = [polar_map[c.split(' ')[0].split('.')[0]] for c in lipids_polar.columns]\n\nall_sample_lists = [gcms_metabolomics.columns, lipids_pos.columns, lipids_neg.columns, lipids_polar.columns, microbiome.columns]\nshared_samples_all_modalities = list(set.intersection(*map(set, all_sample_lists)))\nprint('number shared samples: %i' % len(shared_samples_all_modalities))\nmicrobiome = microbiome[shared_samples_all_modalities]\nlipids_neg = lipids_neg[shared_samples_all_modalities]\nlipids_pos = lipids_pos[shared_samples_all_modalities]\nlipids_polar = lipids_polar[shared_samples_all_modalities]\ngcms_metabolomics = gcms_metabolomics[shared_samples_all_modalities]\nmetadata_noqc_matched = metadata_noqc.loc[shared_samples_all_modalities, :]\nmetadata_noqc_matched['lineage_and_digestive_strategy'] = metadata_noqc_matched['KT_lineage'] + '-' + metadata_noqc_matched['digestive_strategy']\n\nmetadata_noqc_matched.head(5)\n"},{"cell_type":"code","execution_count":3,"metadata":{},"outputs":[{"data":{"text/plain":["Herbivores    50\n","Omnivores     38\n","Carnivores    13\n","Name: diet, dtype: int64"]},"execution_count":3,"metadata":{},"output_type":"execute_result"}],"source":"metadata_noqc_matched.diet.value_counts()"},{"cell_type":"code","execution_count":6,"metadata":{},"outputs":[{"data":{"text/plain":["Primates           30\n","Perissodactyla     30\n","Carnivora          27\n","Afrotheria          8\n","Cetartiodactyla     6\n","Name: KT_lineage, dtype: int64"]},"execution_count":6,"metadata":{},"output_type":"execute_result"}],"source":"metadata_noqc_matched.KT_lineage.value_counts()"},{"cell_type":"code","execution_count":5,"metadata":{},"outputs":[{"data":{"text/plain":["Zebra             16\n","Rhino             13\n","Chimp             10\n","Mandrill           8\n","Elephant           8\n","Coati              7\n","Bear               7\n","Gorilla            6\n","Sheep              5\n","Lion               3\n","CarnivorousCat     3\n","Gibbon             3\n","Wolf               2\n","Tiger              2\n","Leopard            2\n","Lemur              2\n","Goat               1\n","Donkey             1\n","Hyena              1\n","Capuchin           1\n","Name: HostSpecies, dtype: int64"]},"execution_count":5,"metadata":{},"output_type":"execute_result"}],"source":"metadata_noqc_matched.HostSpecies.value_counts()"},{"cell_type":"code","execution_count":4,"metadata":{},"outputs":[{"data":{"text/plain":["Primates-Simplegut         30\n","Perissodactyla-Hindgut     30\n","Carnivora-Simplegut        27\n","Afrotheria-Hindgut          8\n","Cetartiodactyla-Foregut     6\n","Name: lineage_and_digestive_strategy, dtype: int64"]},"execution_count":4,"metadata":{},"output_type":"execute_result"}],"source":"metadata_noqc_matched.lineage_and_digestive_strategy.value_counts()"},{"cell_type":"code","execution_count":9,"metadata":{},"outputs":[],"source":"from sklearn.model_selection import train_test_split\nnp.random.seed(42)\n\nfor i_ in range(10):\n    train, test = train_test_split(metadata_noqc_matched, test_size=0.25, shuffle=True,\n                                   stratify=metadata_noqc_matched[['lineage_and_digestive_strategy']])\n    metadata_noqc_matched.loc[:, 'traintest_%i' % i_] = 'train'\n    metadata_noqc_matched.loc[test.index, 'traintest_%i' % i_] = 'test'\nmetadata_noqc_matched.to_csv('../../data/case-studies/mammalian/matched-data/metadata.tsv', sep='\\t')\n\nfile_save_map = {'gcms':gcms_metabolomics,\n                 'lipid_pos':lipids_pos,\n                 'lipid_neg':lipids_neg,\n                 'lipid_polar':lipids_polar,\n                 'microbiome':microbiome}\nfor name_, table_out in file_save_map.items():\n    # save tables\n    table_out.index = name_ + '_feature_' + table_out.index.astype(str)\n    table_out = Table(table_out.values, table_out.index, table_out.columns)\n    with biom_open('../data/matched-data/%s.biom' % (name_), 'w') as f:\n        table_out.to_hdf5(f, \"filtered-table-cm\")\n"},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":""}],"metadata":{"kernelspec":{"display_name":"joint-rpca-benchmarking-flex","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.9.19"}},"nbformat":4,"nbformat_minor":2}