{"cells":[{"cell_type":"code","execution_count":1,"metadata":{"scrolled":true},"outputs":[],"source":"import os\nimport glob\nimport numpy as np\nimport pandas as pd\nimport mofax as mfx\n\nfrom biom import load_table, Table\nfrom gemelli.rpca import rpca_table_processing\nfrom tqdm.notebook import tqdm\n\nfrom skbio import DistanceMatrix, OrdinationResults\nfrom skbio.stats.composition import closure, clr\nfrom skbio.stats.distance import permanova\n\nfrom scipy.spatial import distance\nfrom sklearn.ensemble import RandomForestClassifier\nfrom sklearn.preprocessing import StandardScaler\nfrom sklearn.metrics import (average_precision_score, auc, roc_auc_score,\n                             balanced_accuracy_score, f1_score)\nfrom sklearn import metrics \n"},{"cell_type":"markdown","metadata":{},"source":"## Import all metric results from file "},{"cell_type":"code","execution_count":2,"metadata":{},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>site_sub_coll</th>\n","      <th>data_type</th>\n","      <th>week_num</th>\n","      <th>date_of_receipt</th>\n","      <th>interval_days</th>\n","      <th>visit_num</th>\n","      <th>Research Project</th>\n","      <th>PDO Number</th>\n","      <th>GSSR IDs</th>\n","      <th>Product</th>\n","      <th>...</th>\n","      <th>train_test_1</th>\n","      <th>train_test_2</th>\n","      <th>train_test_3</th>\n","      <th>train_test_4</th>\n","      <th>train_test_5</th>\n","      <th>train_test_6</th>\n","      <th>train_test_7</th>\n","      <th>train_test_8</th>\n","      <th>train_test_9</th>\n","      <th>train_test_10</th>\n","    </tr>\n","    <tr>\n","      <th>External ID</th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","      <th></th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>CSM67UA2</th>\n","      <td>C3001C15</td>\n","      <td>metabolomics</td>\n","      <td>28.0</td>\n","      <td>2014-09-30</td>\n","      <td>13.0</td>\n","      <td>20</td>\n","      <td>ibdmdb</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>...</td>\n","      <td>train</td>\n","      <td>train</td>\n","      <td>train</td>\n","      <td>train</td>\n","      <td>train</td>\n","      <td>train</td>\n","      <td>train</td>\n","      <td>train</td>\n","      <td>train</td>\n","      <td>train</td>\n","    </tr>\n","    <tr>\n","      <th>CSM67UBN</th>\n","      <td>C3002C17</td>\n","      <td>metabolomics</td>\n","      <td>32.0</td>\n","      <td>2014-12-09</td>\n","      <td>14.0</td>\n","      <td>22</td>\n","      <td>ibdmdb</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>...</td>\n","      <td>train</td>\n","      <td>train</td>\n","      <td>test</td>\n","      <td>train</td>\n","      <td>train</td>\n","      <td>train</td>\n","      <td>train</td>\n","      <td>train</td>\n","      <td>train</td>\n","      <td>test</td>\n","    </tr>\n","    <tr>\n","      <th>CSM67UBZ</th>\n","      <td>C3003C14</td>\n","      <td>metabolomics</td>\n","      <td>26.0</td>\n","      <td>2014-11-06</td>\n","      <td>15.0</td>\n","      <td>19</td>\n","      <td>ibdmdb</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>...</td>\n","      <td>train</td>\n","      <td>train</td>\n","      <td>train</td>\n","      <td>test</td>\n","      <td>test</td>\n","      <td>train</td>\n","      <td>train</td>\n","      <td>test</td>\n","      <td>test</td>\n","      <td>train</td>\n","    </tr>\n","    <tr>\n","      <th>CSM79HLA</th>\n","      <td>C3004C19</td>\n","      <td>metabolomics</td>\n","      <td>36.0</td>\n","      <td>2015-01-20</td>\n","      <td>14.0</td>\n","      <td>25</td>\n","      <td>ibdmdb</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>...</td>\n","      <td>train</td>\n","      <td>test</td>\n","      <td>train</td>\n","      <td>train</td>\n","      <td>test</td>\n","      <td>train</td>\n","      <td>train</td>\n","      <td>train</td>\n","      <td>train</td>\n","      <td>train</td>\n","    </tr>\n","    <tr>\n","      <th>CSM67UEI</th>\n","      <td>C3005C19</td>\n","      <td>metabolomics</td>\n","      <td>36.0</td>\n","      <td>2015-01-29</td>\n","      <td>16.0</td>\n","      <td>25</td>\n","      <td>ibdmdb</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>...</td>\n","      <td>train</td>\n","      <td>train</td>\n","      <td>train</td>\n","      <td>train</td>\n","      <td>train</td>\n","      <td>test</td>\n","      <td>train</td>\n","      <td>train</td>\n","      <td>train</td>\n","      <td>test</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>5 rows Ã— 499 columns</p>\n","</div>"],"text/plain":["            site_sub_coll     data_type  week_num date_of_receipt  \\\n","External ID                                                         \n","CSM67UA2         C3001C15  metabolomics      28.0      2014-09-30   \n","CSM67UBN         C3002C17  metabolomics      32.0      2014-12-09   \n","CSM67UBZ         C3003C14  metabolomics      26.0      2014-11-06   \n","CSM79HLA         C3004C19  metabolomics      36.0      2015-01-20   \n","CSM67UEI         C3005C19  metabolomics      36.0      2015-01-29   \n","\n","             interval_days  visit_num Research Project  PDO Number  GSSR IDs  \\\n","External ID                                                                    \n","CSM67UA2              13.0         20           ibdmdb         NaN       NaN   \n","CSM67UBN              14.0         22           ibdmdb         NaN       NaN   \n","CSM67UBZ              15.0         19           ibdmdb         NaN       NaN   \n","CSM79HLA              14.0         25           ibdmdb         NaN       NaN   \n","CSM67UEI              16.0         25           ibdmdb         NaN       NaN   \n","\n","             Product  ...  train_test_1  train_test_2  train_test_3  \\\n","External ID           ...                                             \n","CSM67UA2         NaN  ...         train         train         train   \n","CSM67UBN         NaN  ...         train         train          test   \n","CSM67UBZ         NaN  ...         train         train         train   \n","CSM79HLA         NaN  ...         train          test         train   \n","CSM67UEI         NaN  ...         train         train         train   \n","\n","             train_test_4  train_test_5  train_test_6  train_test_7  \\\n","External ID                                                           \n","CSM67UA2            train         train         train         train   \n","CSM67UBN            train         train         train         train   \n","CSM67UBZ             test          test         train         train   \n","CSM79HLA            train          test         train         train   \n","CSM67UEI            train         train          test         train   \n","\n","             train_test_8  train_test_9  train_test_10  \n","External ID                                             \n","CSM67UA2            train         train          train  \n","CSM67UBN            train         train           test  \n","CSM67UBZ             test          test          train  \n","CSM79HLA            train         train          train  \n","CSM67UEI            train         train           test  \n","\n","[5 rows x 499 columns]"]},"metadata":{},"output_type":"display_data"}],"source":"max_iterations = 5\nmetadata = pd.read_csv('../../data/simulations/ihmp/sample-metadata-plus-train-tests-last.csv', index_col=0)\ndisplay(metadata.head(5))"},{"cell_type":"code","execution_count":5,"metadata":{},"outputs":[{"data":{"application/vnd.jupyter.widget-view+json":{"model_id":"4ba4ef65a1634e32893c4ce804520cf4","version_major":2,"version_minor":0},"text/plain":["  0%|          | 0/60 [00:00<?, ?it/s]"]},"metadata":{},"output_type":"display_data"}],"source":"ord_res_mofa = {}\n\nfor k in tqdm(glob.glob('../../results/simulations/ihmp/mofa/*factors.model.*')):\n    # save ordination\n    df_ = pd.read_csv(k, index_col=0)\n    df_ = pd.pivot_table(columns='factor', values='value', index='sample', data=df_)\n    # add to metadata for ML\n    df_ = df_[['Factor1','Factor2','Factor3']]\n    df_.columns = ['PC1','PC2','PC3']\n    # save to dict\n    f = int(k.split('/')[-1].split('.')[0])\n    if k.split('/')[-1].split('.')[3] == '':\n        k2 = 'meta_g_taxonomic_profiles'\n    else:\n        k2 = 'meta_g_taxonomic_profiles_' + k.split('/')[-1].split('.')[3]\n    if k2 == 'meta_g_taxonomic_profiles':\n        model = mfx.mofa_model(\"../../results/simulations/ihmp/mofa/%i.model..hdf5TRUE\" % (f))\n    else:\n        model = mfx.mofa_model(\"../../results/simulations/ihmp/mofa/%i.model.%s.hdf5TRUE\" % (f, k2.split('_')[-1]))\n    # project new data\n    test_data_project = pd.read_csv(\"../../data/simulations/ihmp/mofa-reformatted/fold-%i-subset-test-%s-test.tsv.gz\" % (f, k2), sep='\\t', compression='gzip')\n    # MOFAx projects on a single view, they vary greatly so we project on each one to be fair.\n    for view_use in model.views:  \n        test_data_project_view = test_data_project[test_data_project.view == view_use]\n        test_data_project_view_X = pd.pivot_table(columns='sample',\n                                                  index='feature',\n                                                  values='value', data=test_data_project_view)\n        # ensure projection data is ordered the same as the input data\n        test_data_project_view_X = test_data_project_view_X.loc[[int(x) for x in model.features[view_use]], :]\n        # project\n        new_values = np.stack([model.project_data(test_data_project_view_X.values.T, view=view_use, factors=i) for i in range(3)]).T\n        projected_factors = pd.DataFrame(new_values, test_data_project_view_X.columns, [['PC1','PC2','PC3']])\n        projected_factors.columns = ['PC1','PC2','PC3']\n        df_projected_ = pd.concat([df_, projected_factors], axis=0)\n        #df_projected_.to_csv(\"../../results/simulations/ihmp/mofa/fold-%i-%s-projected-results-all.csv\" % (f, view_use))\n        # get distance from ordination\n        Udist = distance.cdist(df_projected_, df_projected_)\n        U_dist_res = DistanceMatrix(Udist, ids=df_projected_.index)\n        # add metdata and save\n        df_projected_ = pd.concat([df_projected_, metadata], axis=1)\n        ord_res_mofa[(f, k2, view_use)] = df_projected_\n    # close model now that we are done\n    model.close()\n"},{"cell_type":"code","execution_count":6,"metadata":{},"outputs":[],"source":"ord_res_mixomics = {}\ndist_res_mixomics = {}\n\nfor k in glob.glob('../../results/simulations/ihmp/mixomics/*factors.model.*'):\n    # save ordination\n    df_ = pd.read_csv(k, index_col=0)\n    k = k.replace('_','')\n    if '..' in k.split('/')[-1]:\n        k2 = 'meta_g_taxonomic_profiles'\n    else:\n        k2 = 'meta_g_taxonomic_profiles_' + k.split('/')[-1].split('.')[3]\n    relabel = {'metabolomics':'HMP2_metabolomics',\n              'metatranscriptomics':'meta_t_ecs',\n              'viromes':'virome_virmap_analysis',\n              'metagenomics':k2}\n    f = int(k.split('/')[-1].split('.')[0])\n    for view_use, view_df in df_.groupby('.id'):\n        view_df = view_df.set_index('subject')[['comp1','comp2','comp3']]\n        view_df.columns = ['PC1','PC2','PC3']\n        # get distance from ordination\n        Udist = distance.cdist(view_df, view_df)\n        U_dist_res = DistanceMatrix(Udist, ids=view_df.index)\n        view_df = pd.concat([view_df, metadata], axis=1)\n        ord_res_mixomics[(f, relabel[view_use])] = view_df\n        dist_res_mixomics[(f, relabel[view_use])] = U_dist_res\n"},{"cell_type":"code","execution_count":7,"metadata":{},"outputs":[],"source":"ord_res_icluster = {}\ndist_res_icluster = {}\n\nfor k in glob.glob('../../results/simulations/ihmp/icluster/*factors.model.*'):\n    # save ordination\n    view_df = pd.read_csv(k, index_col=0)\n    k = k.replace('_','')\n    if '..' in k.split('/')[-1]:\n        k2 = 'meta_g_taxonomic_profiles'\n    else:\n        k2 = 'meta_g_taxonomic_profiles_' + k.split('/')[-1].split('.')[3]\n    f = int(k.split('/')[-1].split('.')[0])\n    view_df.columns = ['PC1','PC2','PC3']\n    # get distance from ordination\n    Udist = distance.cdist(view_df, view_df)\n    U_dist_res = DistanceMatrix(Udist, ids=view_df.index)\n    view_df = pd.concat([view_df, metadata], axis=1)\n    ord_res_icluster[(f, k2)] = view_df\n    dist_res_icluster[(f, k2)] = U_dist_res\n"},{"cell_type":"code","execution_count":10,"metadata":{},"outputs":[],"source":"ord_res_intNMF = {}\ndist_res_intNMF = {}\n\nfor k in glob.glob('../../results/simulations/ihmp/intNMF/*factors.model.*'):\n    # save ordination\n    view_df = pd.read_csv(k, index_col=0)\n    k = k.replace('_','')\n    if '..' in k.split('/')[-1]:\n        k2 = 'meta_g_taxonomic_profiles'\n    else:\n        k2 = 'meta_g_taxonomic_profiles_' + k.split('/')[-1].split('.')[3]\n    f = int(k.split('/')[-1].split('.')[0])\n    view_df.columns = ['PC1','PC2','PC3']\n    # get distance from ordination\n    Udist = distance.cdist(view_df, view_df)\n    U_dist_res = DistanceMatrix(Udist, ids=view_df.index)\n    view_df = pd.concat([view_df, metadata], axis=1)\n    ord_res_intNMF[(f, k2)] = view_df\n    dist_res_intNMF[(f, k2)] = U_dist_res"},{"cell_type":"code","execution_count":11,"metadata":{"scrolled":false},"outputs":[{"data":{"application/vnd.jupyter.widget-view+json":{"model_id":"6d550cfa3a784168ad8c7256bbb9755b","version_major":2,"version_minor":0},"text/plain":["  0%|          | 0/140 [00:00<?, ?it/s]"]},"metadata":{},"output_type":"display_data"}],"source":"max_iterations = 5\n\nord_all_res_joint = {}\ndist_all_res_joint = {}\nfor k in tqdm(glob.glob('../../results/simulations/ihmp/joint-rpca-results/*')):\n    #print('dist-test-%iiter.txt' % max_iterations)\n    if 'dist-test-%iiter-last.txt' % max_iterations in k:\n        dist_all_res_joint[(int(k.split('/')[-1].split('-')[0]),\n                            k.split('/')[-1].split('-')[1])] = DistanceMatrix.read(k)\n    if 'ord-test-%iiter-last.txt' % max_iterations in k:\n        df_ = OrdinationResults.read(k).samples\n        df_.columns = ['PC1','PC2','PC3']\n        df_ = pd.concat([df_, metadata], axis=1)\n        ord_all_res_joint[(int(k.split('/')[-1].split('-')[0]),\n                           k.split('/')[-1].split('-')[1])] = df_\n            "},{"cell_type":"code","execution_count":12,"metadata":{},"outputs":[],"source":"ord_res = {}\ndist_res = {}\nfor k in glob.glob('../../results/simulations/ihmp/rpca/*'):\n    if 'dist' in k:\n        dist_res[(int(k.split('/')[-1].split('-')[0]),\n                           k.split('/')[-1].split('-')[1])] = DistanceMatrix.read(k)\n    if 'ord' in k:\n        ord_res[(int(k.split('/')[-1].split('-')[0]),\n                           k.split('/')[-1].split('-')[1])] = pd.read_csv(k, index_col=0)\n\n        "},{"cell_type":"code","execution_count":13,"metadata":{"scrolled":false},"outputs":[],"source":"ord_res_others = {}\ndist_res_others = {}\nfor k in glob.glob('../../results/simulations/ihmp/all-others/*'):\n    if 'dist' in k:\n        dist_res_others[(int(k.split('/')[-1].split('-')[0]),\n                           k.split('/')[-1].split('-')[1],\n                           k.split('/')[-1].split('-')[2])] = DistanceMatrix.read(k)\n    if 'ord' in k:\n        ord_res_others[(int(k.split('/')[-1].split('-')[0]),\n                           k.split('/')[-1].split('-')[1],\n                           k.split('/')[-1].split('-')[2])] = pd.read_csv(k, index_col=0)\n\n"},{"cell_type":"markdown","metadata":{},"source":"## Run ML on ordinations"},{"cell_type":"code","execution_count":32,"metadata":{},"outputs":[{"data":{"application/vnd.jupyter.widget-view+json":{"model_id":"a3801f03522a406096d52c7da06e77a3","version_major":2,"version_minor":0},"text/plain":["  0%|          | 0/10 [00:00<?, ?it/s]"]},"metadata":{},"output_type":"display_data"}],"source":"# merge all ords. (for MOFA+ we chose the HMP2_metabolomics since it had the best seperation)\nord_res_all = {**{(k[0], k[1], 'Joint-RPCA'):v for k, v in ord_all_res_joint.items()},\n               **{(k[0], k[1], 'iCluster'):v for k, v in ord_res_icluster.items()},\n               **{(k[0], k[1], 'intNMF'):v for k, v in ord_res_intNMF.items()},\n               **{(int(k[0]), k[1], 'mixOmics'):v for k, v in ord_res_mixomics.items()},\n               **{(int(k[0]), k[1], k[2], 'MOFA'):v for k, v in ord_res_mofa.items()},\n               **{(k[0], k[1], 'RPCA'):v for k, v in ord_res.items()},\n              }\naccuracy_score = {}\nf1_score = {}\npermanova_res = {}\napr_scores = {}\nauc_roc_scores = {}\n\ncols_learn = ['PC1','PC2','PC3']\ncovert_map = {'IBD': 1, 'nonIBD': 0}\nclassifier = RandomForestClassifier(n_estimators=500, \n                                    random_state=1010)\n\nfor fold in tqdm(range(10)):\n    fold += 1 # (saved in the dataframe as 1-10)\n    #\n    # for all sparsity levels\n    for dt_name in ['meta_g_taxonomic_profiles_11',\n                    'meta_g_taxonomic_profiles_9',\n                    'meta_g_taxonomic_profiles_7',\n                    'meta_g_taxonomic_profiles_5',\n                    'meta_g_taxonomic_profiles_3',\n                    'meta_g_taxonomic_profiles']:\n        \n        # all metrics\n        for metric_ in ['iCluster','intNMF',\n                        'Joint-RPCA','MOFA','mixOmics']:\n            # make a list to append for the dicts\n            if (dt_name, metric_) not in accuracy_score.keys():\n                accuracy_score[(dt_name, metric_)] = []\n                f1_score[(dt_name, metric_)] = []\n                permanova_res[(dt_name, metric_)] = []\n                apr_scores[(dt_name, metric_)] = []\n                auc_roc_scores[(dt_name, metric_)] = []\n            # MOFA projects on multiple views so we will test them all\n            if metric_ == 'MOFA':\n                for other_tbl_ in  [dt_name, 'virome_virmap_analysis','meta_t_ecs','HMP2_metabolomics']:\n                    # get the ordination data\n                    tbl = ord_res_all[(fold, dt_name, other_tbl_, metric_)].copy()   \n                    tbl = tbl[cols_learn].dropna(subset=cols_learn)\n                    # get labels\n                    metadata_train = metadata[metadata['train_test_%i' % (fold)] == 'train']\n                    metadata_test = metadata[metadata['train_test_%i' % (fold)] == 'test']\n                    metadata_train = metadata_train.loc[list(set(tbl.index) & set(metadata_train.index)), :]\n                    metadata_test = metadata_test.loc[list(set(tbl.index) & set(metadata_test.index)), :]\n                    y_train = list(metadata_train['diagnosis'].values)\n                    y_train = [covert_map[i] for i in y_train]\n                    y_test = list(metadata_test['diagnosis'].values)\n                    y_test = [covert_map[i] for i in y_test]\n                    X_train = tbl.loc[metadata_train.index, :].values\n                    X_test = tbl.loc[metadata_test.index, :].values\n                    # permanova on test data (projection)\n                    dist_tmp = DistanceMatrix(distance.cdist(tbl, tbl), tbl.index)\n                    permanova_res[(dt_name, metric_)].append(permanova(dist_tmp, metadata.loc[dist_tmp.ids, ['diagnosis']].iloc[:, 0], permutations=1)['test statistic'])\n                    # ML\n                    scaler = StandardScaler()\n                    X_train = scaler.fit_transform(X_train)\n                    X_test = scaler.transform(X_test)\n                    classifier.fit(X_train, y_train)\n                    y_score = classifier.predict_proba(X_test)[:, 1]\n                    y_pred = classifier.predict(X_test)\n                    accuracy_score[(dt_name, metric_)].append(balanced_accuracy_score(y_test, y_pred))\n                    f1_score[(dt_name, metric_)].append(metrics.f1_score(y_test, y_pred, average='macro'))\n                    apr_scores[(dt_name, metric_)].append(average_precision_score(y_test, y_score))\n                    auc_roc_scores[(dt_name, metric_)].append(roc_auc_score(y_test, y_score))\n\n            # all others project on a single view\n            else:\n                # get the ordination data\n                tbl = ord_res_all[(fold, dt_name, metric_)].copy()   \n                tbl = tbl[cols_learn].dropna(subset=cols_learn)\n                # get labels\n                metadata_train = metadata[metadata['train_test_%i' % (fold)] == 'train']\n                metadata_test = metadata[metadata['train_test_%i' % (fold)] == 'test']\n                metadata_train = metadata_train.loc[list(set(tbl.index) & set(metadata_train.index)), :]\n                metadata_test = metadata_test.loc[list(set(tbl.index) & set(metadata_test.index)), :]\n                y_train = list(metadata_train['diagnosis'].values)\n                y_train = [covert_map[i] for i in y_train]\n                y_test = list(metadata_test['diagnosis'].values)\n                y_test = [covert_map[i] for i in y_test]\n                X_train = tbl.loc[metadata_train.index, :].values\n                X_test = tbl.loc[metadata_test.index, :].values\n                # permanova on test data (projection)\n                dist_tmp = DistanceMatrix(distance.cdist(tbl, tbl), tbl.index)\n                permanova_res[(dt_name, metric_)].append(permanova(dist_tmp, metadata.loc[dist_tmp.ids, ['diagnosis']].iloc[:, 0], \n                                                                    permutations=1)['test statistic'])\n                # ML\n                scaler = StandardScaler()\n                X_train = scaler.fit_transform(X_train)\n                X_test = scaler.transform(X_test)\n                classifier.fit(X_train, y_train)\n                y_score = classifier.predict_proba(X_test)[:, 1]\n                y_pred = classifier.predict(X_test)\n                accuracy_score[(dt_name, metric_)].append(balanced_accuracy_score(y_test, y_pred))\n                f1_score[(dt_name, metric_)].append(metrics.f1_score(y_test, y_pred, average='macro'))\n                apr_scores[(dt_name, metric_)].append(average_precision_score(y_test, y_score))\n                auc_roc_scores[(dt_name, metric_)].append(roc_auc_score(y_test, y_score)) \n                "},{"cell_type":"code","execution_count":35,"metadata":{},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>dataset</th>\n","      <th>method</th>\n","      <th>fold</th>\n","      <th>f_stat</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>meta_g_taxonomic_profiles_11</td>\n","      <td>iCluster</td>\n","      <td>0</td>\n","      <td>1.482600</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>meta_g_taxonomic_profiles_11</td>\n","      <td>iCluster</td>\n","      <td>1</td>\n","      <td>2.476490</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>meta_g_taxonomic_profiles_11</td>\n","      <td>iCluster</td>\n","      <td>2</td>\n","      <td>1.928308</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>meta_g_taxonomic_profiles_11</td>\n","      <td>iCluster</td>\n","      <td>3</td>\n","      <td>0.258620</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>meta_g_taxonomic_profiles_11</td>\n","      <td>iCluster</td>\n","      <td>4</td>\n","      <td>1.387822</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>"],"text/plain":["                        dataset    method  fold    f_stat\n","0  meta_g_taxonomic_profiles_11  iCluster     0  1.482600\n","1  meta_g_taxonomic_profiles_11  iCluster     1  2.476490\n","2  meta_g_taxonomic_profiles_11  iCluster     2  1.928308\n","3  meta_g_taxonomic_profiles_11  iCluster     3  0.258620\n","4  meta_g_taxonomic_profiles_11  iCluster     4  1.387822"]},"execution_count":35,"metadata":{},"output_type":"execute_result"}],"source":"permanova_all_df = pd.DataFrame(dict([ (k,pd.Series(v)) for k,v in permanova_res.items() ])).T.stack().reset_index().dropna(axis=1)\npermanova_all_df.columns = ['dataset','method','fold','f_stat']\npermanova_all_df['method'] = permanova_all_df.method.replace('aitchison','Aitchison').replace('braycurtis','Bray-Curtis')\npermanova_all_df.to_csv('../../data/simulations/benchmarks/permanova-test-%iiter-last.csv' % max_iterations)\npermanova_all_df.head(5)"},{"cell_type":"code","execution_count":36,"metadata":{},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>dataset</th>\n","      <th>method</th>\n","      <th>fold</th>\n","      <th>apr</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>meta_g_taxonomic_profiles_11</td>\n","      <td>iCluster</td>\n","      <td>0</td>\n","      <td>0.796900</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>meta_g_taxonomic_profiles_11</td>\n","      <td>iCluster</td>\n","      <td>1</td>\n","      <td>0.789058</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>meta_g_taxonomic_profiles_11</td>\n","      <td>iCluster</td>\n","      <td>2</td>\n","      <td>0.840364</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>meta_g_taxonomic_profiles_11</td>\n","      <td>iCluster</td>\n","      <td>3</td>\n","      <td>0.662699</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>meta_g_taxonomic_profiles_11</td>\n","      <td>iCluster</td>\n","      <td>4</td>\n","      <td>0.836612</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>"],"text/plain":["                        dataset    method  fold       apr\n","0  meta_g_taxonomic_profiles_11  iCluster     0  0.796900\n","1  meta_g_taxonomic_profiles_11  iCluster     1  0.789058\n","2  meta_g_taxonomic_profiles_11  iCluster     2  0.840364\n","3  meta_g_taxonomic_profiles_11  iCluster     3  0.662699\n","4  meta_g_taxonomic_profiles_11  iCluster     4  0.836612"]},"execution_count":36,"metadata":{},"output_type":"execute_result"}],"source":"apr_all_df = pd.DataFrame(dict([ (k,pd.Series(v)) for k,v in apr_scores.items() ])).T.stack().reset_index().dropna(axis=1)\napr_all_df.columns = ['dataset','method','fold','apr']\napr_all_df['method'] = apr_all_df.method.replace('aitchison','Aitchison').replace('braycurtis','Bray-Curtis')\napr_all_df.to_csv('../../data/simulations/benchmarks/apr-test-%iiter-last.csv' % max_iterations)\napr_all_df.head(5)"},{"cell_type":"code","execution_count":37,"metadata":{},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>dataset</th>\n","      <th>method</th>\n","      <th>fold</th>\n","      <th>roc_auc</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>meta_g_taxonomic_profiles_11</td>\n","      <td>iCluster</td>\n","      <td>0</td>\n","      <td>0.533333</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>meta_g_taxonomic_profiles_11</td>\n","      <td>iCluster</td>\n","      <td>1</td>\n","      <td>0.725000</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>meta_g_taxonomic_profiles_11</td>\n","      <td>iCluster</td>\n","      <td>2</td>\n","      <td>0.600000</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>meta_g_taxonomic_profiles_11</td>\n","      <td>iCluster</td>\n","      <td>3</td>\n","      <td>0.450000</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>meta_g_taxonomic_profiles_11</td>\n","      <td>iCluster</td>\n","      <td>4</td>\n","      <td>0.633333</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>"],"text/plain":["                        dataset    method  fold   roc_auc\n","0  meta_g_taxonomic_profiles_11  iCluster     0  0.533333\n","1  meta_g_taxonomic_profiles_11  iCluster     1  0.725000\n","2  meta_g_taxonomic_profiles_11  iCluster     2  0.600000\n","3  meta_g_taxonomic_profiles_11  iCluster     3  0.450000\n","4  meta_g_taxonomic_profiles_11  iCluster     4  0.633333"]},"execution_count":37,"metadata":{},"output_type":"execute_result"}],"source":"roc_all_df = pd.DataFrame(dict([ (k,pd.Series(v)) for k,v in auc_roc_scores.items() ])).T.stack().reset_index().dropna(axis=1)\nroc_all_df.columns = ['dataset','method','fold','roc_auc']\nroc_all_df['method'] = roc_all_df.method.replace('aitchison','Aitchison').replace('braycurtis','Bray-Curtis')\nroc_all_df.to_csv('../../data/simulations/benchmarks/roc-auc-test-%iiter-last.csv' % max_iterations)\nroc_all_df.head(5)"}],"metadata":{"kernelspec":{"display_name":"joint-rpca-benchmarking-flex","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.9.19"}},"nbformat":4,"nbformat_minor":4}